packages/backend/server/src/plugins/copilot/context/session.ts

```diff
@@ -0,0 +1,236 @@
+    return this.db.$transaction(async tx => {
+      await tx.$executeRaw`
+        INSERT INTO "ai_context_embeddings"
+        ("id", "context_id", "file_id", "chunk", "content", "embedding", "updated_at") VALUES ${values}
+        ON CONFLICT (context_id, file_id, chunk) DO UPDATE SET
+        content = EXCLUDED.content, embedding = EXCLUDED.embedding, updated_at = excluded.updated_at;
+      `;
+      await this.saveFileRecord(
+        fileId,
+        file => ({
+          ...(file as ContextFile),
+          status: ContextFileStatus.finished,
+        }),
+        tx
+      );
+      return fileId;
+    });
+  }
+
+  private readStream(
+    readable: Readable,
+    maxSize = 50 * OneMB
+  ): Promise<Buffer<ArrayBuffer>> {
+    return new Promise<Buffer<ArrayBuffer>>((resolve, reject) => {
+      const chunks: Uint8Array[] = [];
+      let totalSize = 0;
+
+      readable.on('data', chunk => {
+        totalSize += chunk.length;
+        if (totalSize > maxSize) {
+          reject(new BlobQuotaExceeded());
+          readable.destroy(new BlobQuotaExceeded());
+          return;
+        }
+        chunks.push(chunk);
+      });
+
+      readable.on('end', () => {
+        resolve(Buffer.concat(chunks, totalSize));
+      });
+
+      readable.on('error', err => {
+        reject(err);
+      });
+    });
+  }
+
+  async addDocRecord(docId: string) {
+    if (!this.config.docs.includes(docId)) {
+      this.config.docs.push(docId);
+      await this.save();
+    }
+    return this.config.docs.length;
+  }
+
+  async removeDocRecord(docId: string) {
+    const index = this.config.docs.indexOf(docId);
+    if (index >= 0) {
+      this.config.docs.splice(index, 1);
+      await this.save();
+      return true;
+    }
+    return false;
+  }
+
+  async addStream(
+    readable: Readable,
+    name: string,
+    blobId: string,
+    signal?: AbortSignal
+  ): Promise<string | undefined> {
+    if (signal?.aborted) return;
+    const buffer = await this.readStream(readable, 50 * OneMB);
+    const file = new File([buffer], name);
+    return await this.add(file, blobId, signal);
+  }
+
+  async add(
+    file: File,
+    blobId: string,
+    signal?: AbortSignal
+  ): Promise<string | undefined> {
+    const embeddings = await this.client.getFileEmbeddings(file, signal);
+    if (embeddings && !signal?.aborted) {
+      return await this.insertEmbeddings(file.name, blobId, embeddings);
+    }
+    return undefined;
+  }
+
+  async remove(fileId: string) {
+    return await this.db.$transaction(async tx => {
+      const ret = await tx.aiContextEmbedding.deleteMany({
+        where: { contextId: this.contextId, fileId },
+      });
+      this.config.files = this.config.files.filter(f => f.id !== fileId);
+      await this.save(tx);
+      return ret.count > 0;
+    });
+  }
+
+  async match(
+    content: string,
+    topK: number = 5,
+    signal?: AbortSignal
+  ): Promise<FileChunkSimilarity[]> {
```

Comment:
Rename the function name to `matchFile`? There may be a `matchDoc` function in workspace embedding feature.